index.js

```diff
@@ -105,6 +105,12 @@ module.exports = function (na) {
     return hash
   }
 
+  exports.crypto_generichash = function (ptxt, length, key) {
+    var hash = Z(length || na.crypto_generichash_BYTES);
```

Comment:
If length is not specified, then `length` should become `key`?  As in, if someone wants to specify `key` they must also specify length?

Reply:
And It looks like there is also `crypto_generichash_KEYBYTES` to make use of here as well:

https://github.com/sodium-friends/sodium-native#generic-hashing

Reply:
> If length is not specified, then length should become key? As in, if someone wants to specify key they must also specify length?

That seems useful but I think the goal of this adapter library is to adhere to the native API, which doesn't support something like that (I don't think). You'd have to provide all the parameters if you want to use the key. I could be wrong -- I'll leave that decision up to the maintainer.

Reply:
I just noticed the readme indicates `sodium-friends needs a PR to add key conversions`.  Looking at the code, it appears to me that there is no mechanism to convert a `key` string into the `unsigned char *key` and `unsigned long long key_length` components that feed this: https://github.com/sodium-friends/sodium-native/blob/master/src/crypto_generichash_wrap.cc#L38

@dlongley did you discover that it is possible to pass in a key as a string without a patch to `sodium-native`?

It's unclear to me whether the length of `key` should be tested against `crypto_generichash_KEYBYTES_MIN` and `crypto_generichash_KEYBYTES_MAX` at the wrapper layer or let it potentially fail in the native code.  Seems like failing while still in JS land would be better since those constants are exposed.  But, since I guess it would throw a `TypeError` (or the C equivaltent), efficient failure is probably a non-issue.

Reply:
> did you discover that it is possible to pass in a key as a string without a patch to `sodium-native`?

The key should and must be a buffer, not a string.